File: src/main/scala/com/exasol/cloudetl/kafka/KafkaTopicDataImporter.scala
package com.exasol.cloudetl.kafka

import com.exasol.ExaIterator
import com.exasol.ExaMetadata
import com.exasol.cloudetl.kafka.consumer.KafkaRecordConsumer

import com.typesafe.scalalogging.LazyLogging

/**
 * This object is referenced from the UDF script that imports data from
 * a Kafka topic into an Exasol table.
 */
object KafkaTopicDataImporter extends LazyLogging {

  /**
   * Consumes Kafka topic records and emits them into an Exasol table.
   *
   * The function is called for each Kafka topic partition in parallel.
   * It polls all the record offsets that have not been consumed before
   * and emits them to the Exasol table.
   *
   * Together with the record data, it emits the partition id and record
   * offset as metadata.
   */
  def run(metadata: ExaMetadata, iterator: ExaIterator): Unit = {
    val kafkaProperties = KafkaConsumerProperties(iterator.getString(0), metadata)
    val partitionId = iterator.getInteger(1)
    val partitionNextOffset = iterator.getLong(2) + 1L
    val outputColumnCount = metadata.getOutputColumnCount().toInt
    val outputColumnTypes: Seq[Class[_]] = (0 until outputColumnCount).map(x => metadata.getOutputColumnType(x))
    val nodeId = metadata.getNodeId()
    val vmId = metadata.getVmId()
    logger.info(
      s"Starting Kafka consumer for partition '$partitionId' at next offset " +
        s"'$partitionNextOffset' for node '$nodeId' and vm '$vmId'."
    )
    new KafkaRecordConsumer(
      kafkaProperties,
      partitionId,
      partitionNextOffset,
      outputColumnTypes,
      outputColumnCount,
      nodeId,
      vmId
    ).emit(iterator)
  }

}

File: src/main/java/com/exasol/cloudetl/kafka/KafkaTopicDataImporter.java
package com.exasol.cloudetl.kafka;

import java.util.ArrayList;
import java.util.List;

import com.exasol.ExaIterator;
import com.exasol.ExaMetadata;
import com.exasol.cloudetl.kafka.consumer.KafkaRecordConsumer;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public final class KafkaTopicDataImporter {
    private static final Logger LOGGER = LoggerFactory.getLogger(KafkaTopicDataImporter.class);
    private static final int PARTITION_ID_INDEX = 1;
    private static final int MAXIMUM_OFFSET_INDEX = 2;

    private KafkaTopicDataImporter() {
    }

    public static void run(final ExaMetadata metadata, final ExaIterator iterator) throws Exception {
        final KafkaConsumerProperties kafkaProperties = KafkaConsumerProperties.apply(
                iterator.getString(KafkaConnectorConstants.KEY_VALUE_PROPERTIES_INDEX), metadata);
        final int partitionId = iterator.getInteger(PARTITION_ID_INDEX);
        final long partitionNextOffset = iterator.getLong(MAXIMUM_OFFSET_INDEX) + 1L;
        final int outputColumnCount = (int) metadata.getOutputColumnCount();
        final List<Class<?>> outputColumnTypes = new ArrayList<>();
        for (int columnIndex = 0; columnIndex < outputColumnCount; columnIndex++) {
            outputColumnTypes.add(metadata.getOutputColumnType(columnIndex));
        }
        final long nodeId = metadata.getNodeId();
        final String vmId = metadata.getVmId();
        LOGGER.info("Starting Kafka consumer for partition '{}' at next offset '{}' for node '{}' and vm '{}'.",
                partitionId, partitionNextOffset, nodeId, vmId);
        new KafkaRecordConsumer(kafkaProperties, partitionId, partitionNextOffset, ScalaCollections.seq(outputColumnTypes),
                outputColumnCount, nodeId, vmId).emit(iterator);
    }
}
